exporter/lokiexporter/next_exporter.go

```diff
@@ -0,0 +1,176 @@
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/lokiexporter/internal/third_party/loki/logproto"
+)
+
+type nextLokiExporter struct {
+	config   *Config
+	settings component.TelemetrySettings
+	client   *http.Client
+	wg       sync.WaitGroup
+}
+
+func newNextExporter(config *Config, settings component.TelemetrySettings) *nextLokiExporter {
+	settings.Logger.Info("using the new Loki exporter")
+
+	return &nextLokiExporter{
+		config:   config,
+		settings: settings,
+	}
+}
+
+func (l *nextLokiExporter) pushLogData(ctx context.Context, ld plog.Logs) error {
+	pushReq := logDataToLoki(l.settings.Logger, ld)
+	if len(pushReq.Streams) == 0 {
+		return consumererror.NewPermanent(fmt.Errorf("failed to transform logs into Loki log streams"))
+	}
+
+	buf, err := encode(pushReq)
+	if err != nil {
+		return consumererror.NewPermanent(err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", l.config.HTTPClientSettings.Endpoint, bytes.NewReader(buf))
+	if err != nil {
+		return consumererror.NewPermanent(err)
+	}
+
+	for k, v := range l.config.HTTPClientSettings.Headers {
+		req.Header.Set(k, v)
+	}
+	req.Header.Set("Content-Type", "application/x-protobuf")
+
+	resp, err := l.client.Do(req)
+	if err != nil {
+		return consumererror.NewLogs(err, ld)
+	}
+
+	defer func() {
+		_, _ = io.Copy(io.Discard, resp.Body)
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		scanner := bufio.NewScanner(io.LimitReader(resp.Body, maxErrMsgLen))
+		line := ""
+		if scanner.Scan() {
+			line = scanner.Text()
+		}
+		err = fmt.Errorf("HTTP %d %q: %s", resp.StatusCode, http.StatusText(resp.StatusCode), line)
+		return consumererror.NewLogs(err, ld)
+	}
+
+	return nil
+}
+
+func (l *nextLokiExporter) start(_ context.Context, host component.Host) (err error) {
+	client, err := l.config.HTTPClientSettings.ToClient(host, l.settings)
+	if err != nil {
+		return err
+	}
+
+	l.client = client
+
+	return nil
+}
+
+func (l *nextLokiExporter) stop(context.Context) (err error) {
+	l.wg.Wait()
+	return nil
+}
+
+func logDataToLoki(logger *zap.Logger, ld plog.Logs) (pr *logproto.PushRequest) {
+	var errs error
+
+	streams := make(map[string]*logproto.Stream)
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		ills := rls.At(i).ScopeLogs()
+		resource := rls.At(i).Resource()
+		for j := 0; j < ills.Len(); j++ {
+			logs := ills.At(j).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				log := logs.At(k)
+
+				mergedLabels := convertAttributesAndMerge(log.Attributes(), resource.Attributes())
+				// remove the attributes that were promoted to labels
+				removeAttributes(log.Attributes(), mergedLabels)
```

Comment:
Looked at the implementation of this, it is faster to use `RemoveIf` instead of multiple remove calls.